File: SIC_Backend/Repositories/DevicesRepository.cs
using Microsoft.EntityFrameworkCore;
using SIC_Backend.Data;
using SIC_Backend.Data.DTOs;
using SIC_Backend.Data.Models;

namespace SIC_Backend.Repositories
{
    public class DevicesRepository(ApplicationDbContext dbContext) : IDevicesRepository
    {
        public async Task<DeviceDTO> GetDeviceByIdAsync(int id)
        {
            var device = await dbContext.Devices.FindAsync(id);

            if (device == null)
            {
                //throw new NotFoundException("Device not found");
            }

            return DeviceDTO.FromDevice(device!);
        }

        public IEnumerable<DeviceDTO> GetDevicesByUserId(string userId)
        {
            return dbContext.Devices.Where(d => d.UserId == userId)
                .Select(DeviceDTO.FromDevice)
                .ToList();
        }

        public IEnumerable<DeviceDTO> GetDevicesByPlaceId(int placeId)
        {
            return dbContext.Devices.Where(d => d.PlaceId == placeId)
                .Select(DeviceDTO.FromDevice)
                .ToList();
        }

        public IEnumerable<DeviceDTO> GetDevicesBySectionId(int sectionId)
        {
            return dbContext.Devices.Where(d => d.SectionId == sectionId)
                .Select(DeviceDTO.FromDevice)
                .ToList();
        }

        public async Task CreateDeviceAsync(CreateDeviceModel model)
        {
            var device = Device.FromCreateModel(model);

            dbContext.Devices.Add(device);
            
            await dbContext.SaveChangesAsync();
        }

        public async Task UpdateDeviceAsync(int id, UpdateDeviceModel model)
        {
            var device = await dbContext.Devices.FindAsync(id);

            if (device == null)
            {
                //throw new NotFoundException("Device not found");
            }

            device!.Name = model.Name;
            device.PlaceId = model.PlaceId;
            device.SectionId = model.SectionId;

            await dbContext.SaveChangesAsync();
        }

        public async Task DeleteDeviceAsync(int id)
        {
            var device = await dbContext.Devices.FindAsync(id);

            if (device == null)
            {
                //throw new NotFoundException("Device not found");
            }

            dbContext.Devices.Remove(device!);
        }
    }
}

File: sic_home/lib/repositories/devices_repository.dart
import 'package:dio/dio.dart';
import 'package:sic_home/apis/devices_api.dart';
import 'package:sic_home/models/create_device_model.dart';
import 'package:sic_home/models/device.dart';
import 'package:sic_home/models/update_device_model.dart';
import 'package:sic_home/repositories/interfaces/i_devices_repository.dart';

class DevicesRepository implements IDevicesRepository {
  final DevicesApi _devicesApi = DevicesApi(Dio());

  @override
  Future<int> createDevice(CreateDeviceModel device) async {
    return await _devicesApi.createDevice(device);
  }

  @override
  Future<void> deleteDevice(int id) async {
    return await _devicesApi.deleteDevice(id);
  }

  @override
  Future<Device> getDeviceById(int id) async {
    return await _devicesApi.getDevice(id);
  }

  @override
  Future<List<Device>> getDevicesByPlace(int placeId) async {
    return await _devicesApi.getDevicesByPlace(placeId);
  }

  @override
  Future<List<Device>> getDevicesBySection(int sectionId) async {
    return await _devicesApi.getDevicesBySection(sectionId);
  }

  @override
  Future<List<Device>> getDevicesByUser(String userId) async {
    return await _devicesApi.getDevicesByUser(userId);
  }

  @override
  Future<void> updateDevice(int id, UpdateDeviceModel device) async {
    return await _devicesApi.updateDevice(id, device);
  }
}
